File: rust/clikit/src/diagnostic.rs
//! [`Diagnostic`]: one entry of `errors` or `caveats` — `result-record.schema.json`'s
//! `$defs/error`, shared by both arrays. Which array it sits in, not
//! anything on the value itself, says whether it's a reason the outcome
//! isn't plain success or a qualification on a usable one.

use serde::Serialize;
use serde_json::{Map, Value};

use crate::error::ClikitError;
use crate::triage::Triage;
use crate::validate;

/// One structured, actionable statement — an error or a caveat, depending
/// on which array a [`crate::ResultRecordBuilder`] call places it in.
#[derive(Debug, Clone, Serialize)]
pub struct Diagnostic {
    /// Stable machine identity. Its first segment is the status class this
    /// diagnostic belongs to (`<class>.<domain>.<condition>`); `<class>.clikit.*`
    /// is reserved to this library.
    pub code: String,
    /// What went wrong, in the caller's terms, as one line. No stack trace,
    /// no captured output, no repetition of `code`.
    pub message: String,
    /// What the caller does next. Required — a diagnostic without one is
    /// not finished.
    pub triage: Triage,
    /// Bounded structured detail for this diagnostic. Omitted when there is
    /// none, never `{}`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<Map<String, Value>>,
}

impl Diagnostic {
    /// Starts a diagnostic. `code` is validated at
    /// [`crate::ResultRecordBuilder::build`] time, once the record's status
    /// (and hence which class prefix `code` must carry) is known.
    #[must_use]
    pub fn new(code: impl Into<String>, message: impl Into<String>, triage: Triage) -> Self {
        Diagnostic {
            code: code.into(),
            message: message.into(),
            triage,
            context: None,
        }
    }

    /// Adds one `context` entry. A key written twice keeps the last value.
    #[must_use]
    pub fn context(mut self, key: impl Into<String>, value: impl Into<Value>) -> Self {
        self.context
            .get_or_insert_with(Map::new)
            .insert(key.into(), value.into());
        self
    }

    /// Validates everything about this diagnostic except its class prefix,
    /// which only the containing record's status (and which array it sits
    /// in) can decide.
    pub(crate) fn validate(&self) -> Result<(), ClikitError> {
        validate::validate_code(&self.code)?;
        validate::validate_line("message", &self.message, 4096)?;
        self.triage.validate()?;
        if let Some(context) = &self.context {
            if context.is_empty() || context.len() > 32 {
                return Err(ClikitError::TooMany {
                    field: "context",
                    max: 32,
                    actual: context.len(),
                });
            }
            for key in context.keys() {
                validate::validate_member_key("context key", key)?;
            }
        }
        Ok(())
    }
}

File: go/clikit/diagnostic.go
package clikit

import "fmt"

// Diagnostic is one structured, actionable statement of why a result is not
// plain success: an error or a caveat, the two share this shape. `Triage` is
// always required - a diagnostic that does not say what to do next is not
// finished.
type Diagnostic struct {
	Code    string         `json:"code"`
	Message string         `json:"message"`
	Triage  Triage         `json:"triage"`
	Context map[string]any `json:"context,omitempty"`
}

// NewError builds an errors-array member. code's class (its first segment)
// must be one of the nine failure classes - never `caveats`.
func NewError(code, message string, triage Triage, context map[string]any) (Diagnostic, error) {
	class, ok := diagnosticClass(code)
	if !ok {
		return Diagnostic{}, fmt.Errorf("clikit: malformed diagnostic code %q", code)
	}
	if class == string(StatusCaveats) {
		return Diagnostic{}, fmt.Errorf("clikit: error code %q must not be in the caveats class", code)
	}
	return newDiagnostic(code, message, triage, context)
}

// NewCaveat builds a caveats-array member. code's class must be `caveats` -
// an error can never be filed as a caveat.
func NewCaveat(code, message string, triage Triage, context map[string]any) (Diagnostic, error) {
	class, ok := diagnosticClass(code)
	if !ok {
		return Diagnostic{}, fmt.Errorf("clikit: malformed diagnostic code %q", code)
	}
	if class != string(StatusCaveats) {
		return Diagnostic{}, fmt.Errorf("clikit: caveat code %q must be in the caveats class", code)
	}
	return newDiagnostic(code, message, triage, context)
}

func newDiagnostic(code, message string, triage Triage, context map[string]any) (Diagnostic, error) {
	if !isLine(message) {
		return Diagnostic{}, fmt.Errorf("clikit: invalid diagnostic message %q", message)
	}
	if err := triage.validate(); err != nil {
		return Diagnostic{}, err
	}
	if len(context) > 32 {
		return Diagnostic{}, fmt.Errorf("clikit: diagnostic context has %d members, max 32", len(context))
	}
	for k := range context {
		if !isDataKey(k) {
			return Diagnostic{}, fmt.Errorf("clikit: invalid context key %q", k)
		}
	}
	return Diagnostic{Code: code, Message: message, Triage: triage, Context: context}, nil
}

// validateAsError re-checks d as it must appear inside `errors`: well-formed
// and not in the caveats class. Used by Result.Validate on a record built by
// hand rather than through NewError.
func (d Diagnostic) validateAsError() error {
	class, ok := diagnosticClass(d.Code)
	if !ok {
		return fmt.Errorf("clikit: malformed diagnostic code %q", d.Code)
	}
	if class == string(StatusCaveats) {
		return fmt.Errorf("clikit: error code %q must not be in the caveats class", d.Code)
	}
	return d.validateShape()
}

// validateAsCaveat re-checks d as it must appear inside `caveats`.
func (d Diagnostic) validateAsCaveat() error {
	class, ok := diagnosticClass(d.Code)
	if !ok {
		return fmt.Errorf("clikit: malformed diagnostic code %q", d.Code)
	}
	if class != string(StatusCaveats) {
		return fmt.Errorf("clikit: caveat code %q must be in the caveats class", d.Code)
	}
	return d.validateShape()
}

func (d Diagnostic) validateShape() error {
	if !isLine(d.Message) {
		return fmt.Errorf("clikit: invalid diagnostic message %q", d.Message)
	}
	if err := d.Triage.validate(); err != nil {
		return err
	}
	if len(d.Context) > 32 {
		return fmt.Errorf("clikit: diagnostic context has %d members, max 32", len(d.Context))
	}
	for k := range d.Context {
		if !isDataKey(k) {
			return fmt.Errorf("clikit: invalid context key %q", k)
		}
	}
	return nil
}
